Rewrite walkable items into the sector data and add interactions between them

walkawayy opened the pull request:
#### Checklist

- [X] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description
Rewrite of movable blocks and sliding pillars to support floor and ceiling height functions instead of altering room geometry directly. Walkable items are placed into their respective pit `SECTOR` and checked in the floor and ceiling height functions instead of relying on level triggers (NOTE, is this a possible "regression"? Trick bridges could still be placed as statics with no collision and the actual bridge objects are transparent. This is also how builders get around having multiple bridges in a level look different).

Walkables should now all properly interact with each other. Examples include:
* Pulling/pushing movable blocks on trapdoors, drawbridges, and sliding pillars
* Movable blocks properly stacking on each other and other walkables
* Stacks of movable blocks falling when on a trapdoor or drawbridge that opens
* Movable blocks will raise/lower in lifts

TR1 and TR2 test levels  are located in pushblocks-bridges-trapdoors.zip.

I made a checklist of the progress:

- [X] Add floor height and ceiling height functions for movable block
- [X] Add floor height and ceiling height functions for sliding pillar
- [X] Add walkable items to their respective pit `SECTOR` data 
- [X] Movable blocks fall when resting on an opened trapdoor or drawbridge
- [X] Properly save/load movable block item data
- [X] Remove calls to `Room_AlterFloorHeight`
- [x] Set box blocking so enemies respect the new movable blocks and sliding pillars
- [x] Improve camera behavior around movable blocks
- [x] Improve camera behavior around sliding pillars
- [X] Remove and re-add walkable items on flipmap
- [X] Movable blocks land on sliding pillars and 
- [x] Test TR1 test level in `pushblocks-bridges-trapdoors` for bugs
- [x] Port movable block to TRX
- [x] Test TR2 test level in `pushblocks-bridges-trapdoors` for bugs
- [x] Add TR2 lift support
- [x] Fix using blocks as ladders (also fixes OG bug where Lara hang climbs)
- [x] Collapsible floor tiles (falling blocks)

rr- commented on src/libtrx/game/rooms/geometry.c:
```diff
@@ -221,12 +232,12 @@ int16_t Room_GetHeight(
 int16_t Room_GetHeightEx(
     const SECTOR *sector, const int32_t x, const int32_t y, const int32_t z,
-    const bool fix_tilts)
+    const bool fix_tilts, const int16_t ignore_item_num)
```
I don't think `ignore_item_num` is used anywhere?

walkawayy replied:
It's used exactly once in `src/tr1/game/objects/traps/movable_block.c`'s `M_Control`. It's needed to check different statuses of a falling block.

```c
int32_t under_block_height = Room_GetHeightEx(
    sector, item->pos.x, item->pos.y, item->pos.z, false, item_num);
```

rr- commented on src/libtrx/game/anims/frames.c:
```diff
@@ -178,12 +178,14 @@ void Anim_LoadFrames(const int16_t *data, const int32_t data_length)
             continue;
         }
 
+        cur_obj->anim_count++;
```
Worth landing this and stuff related to this as its own commit

walkawayy replied:
I ended up not even needing this. It could be useful in the future, but for now I'll remove it.

rr- commented on src/libtrx/game/game_buf.c:
```diff
@@ -3,26 +3,36 @@
 {
     const size_t aligned_size = (alloc_size + 3) & ~3;
-    return Memory_ArenaAlloc(&m_Allocator, aligned_size);
+    return Memory_ArenaAlloc(&m_Allocator[buffer], aligned_size);
```
Worth making a separate PR. LMK if you'd like me to do that

walkawayy replied:
Yes please. I will make a few smaller PRs as well that includes prep work for this walkable rewrite. Collision code and some movable block cleanup.

rr- commented on src/libtrx/game/collision.c:
```diff
@@ -5,17 +5,69 @@
+    int16_t room_num)
+{
+    const int32_t y = y_pos - obj_height;
+    const int32_t y_top = y - 160;
```
What does 160 signify here?

Additionally, can we put a generic comment on what this function does in very generic terms? (I know it's a nicely refactored stuff from GetCollisionInfo but I must say I don't quite understand the original function.)

walkawayy replied:
It's the same as `const int32_t y_top = y - 160;` in `Collide_GetCollisionInfo`. It's some kind of Core hard coded number to create a collision box slightly above Lara's head.

Looks like TEN calls it `LARA_HEADROOM` with a comment that explains it pretty well.

walkawayy replied:
I think this change and `M_FillSide` can be landed in a separate PR. Then I can add in the `sim_wall` stuff for this PR.

EDIT: done

rr- commented on src/libtrx/game/items/col.c:
```diff
@@ -108,3 +108,38 @@ const BOUNDS_16 *Item_GetBoundsAccurate(const ITEM *const item)
     return result;
 }
+
+BOUNDS_16 Item_RotateBounds(const ITEM *const item, int16_t rot_y)
```
`rot_y` can be `const`

walkawayy replied:
I might need a clarification from @lahm86 on this one. I originally used this `Item_RotateBounds` for both trapdoors and drawbridges. In trview I noticed though that their y rots seemed to be 180 degrees off from each other. However, for some reason this same function works for both. I can `M_RotateBounds` from each object's .c file, but I am kind of confused on why the same rotation function works for both. I thought drawbridges would need `DEG_90` and `-DEG_90` flipped.

lahm86 replied:
Hmm, what makes you say that about deg 90? I'm not sure I'm following

walkawayy replied:
Resolving this as Lahm explained it:
`Yes, I think your rotation function is fine. The data is just inconsistent, so while the position angles are different, the frame rots make them actually face the same way`

I should be able to use the general rotation function and remove the object specific versions. I will make that change.

rr- commented on src/libtrx/game/items/col.c:
```diff
@@ -108,3 +108,38 @@ const BOUNDS_16 *Item_GetBoundsAccurate(const ITEM *const item)
+BOUNDS_16 Item_RotateBounds(const ITEM *const item, int16_t rot_y)
+{
+    const BOUNDS_16 *bounds = &Item_GetBestFrame(item)->bounds;
+    BOUNDS_16 rot_bounds = { 0 };
```
Zero-initializers can be left as just `{}` thanks to C2x

rr- commented on src/libtrx/game/items/common.c:
```diff
@@ -8,7 +8,11 @@
 #include "utils.h"
+#include "vector.h"
+
+#include <stdlib.h>
```
Stray include?

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,167 @@
+    WALKABLE *node = GameBuf_Alloc(sizeof(WALKABLE), GBUF_WALKABLES);
+    node->item_num = item_num;
+    node->pos = pos;
+    node->next = NULL;
```
We use `nullptr` rather than `NULL` (here and in other places).

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,167 @@
+    LOG_DEBUG(
+        "Add %d; xyz: %d %d %d; floor.height: %d; ceiling.height: %d", item_num,
+        pos.x, pos.y, pos.z, sector->floor.height, sector->ceiling.height);
+    WALKABLE *node = GameBuf_Alloc(sizeof(WALKABLE), GBUF_WALKABLES);
```
`const` and stuff, here in other pointers (I'd appreciate a general recheck prior to final merge – I'll avoid pinpointing every case to avoid noise in the CR)

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,167 @@
+            "Trapdoor item %d OG pos: %d %d %d", item_num, item->pos.x,
+            item->pos.y, item->pos.z);
+        VECTOR *trapdoor_pos = Vector_Create(sizeof(XYZ_32));
+        Trapdoor_GetSectorPositions(item, trapdoor_pos);
```
WDYT about making this function a part of the `OBJECT` structure, as a function pointer, similar to for example `draw` or `control`? This way we don't have to hardcode object IDs here.

walkawayy replied:
Great idea! Done. Added `add_walkable_func` to objects and added `M_AddWalkable` to movable blocks, sliding pillars, trapdoors, bridges, and drawbridges.

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,167 @@
+        }
+        walkables = &(*walkables)->next;
+    }
+    if (!node) {
```
Rather than `(!node)` we do null checks very explicitly `(node == nullptr)`. (Here and other places)

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,167 @@
+void Walkable_Shutdown(void)
+{
+    const int32_t room_count = Room_GetCount();
+    for (int32_t r = 0; r < room_count; ++r) {
```
In general post-incrementation is favored over pre-incrementation.

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,167 @@
+{
+    const int32_t room_count = Room_GetCount();
+    for (int32_t r = 0; r < room_count; ++r) {
+        ROOM *room = (ROOM *)Room_Get(r);
```
No need for this cast.

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,167 @@
+        SECTOR *sec = room->sectors;
+        for (int32_t z = 0; z < nz; ++z)
+            for (int32_t x = 0; x < nx; ++x, ++sec)
+                sec->walkable = NULL;
```
Missing braces.

rr- commented on src/libtrx/game/objects/general/drawbridge.c:
```diff
@@ -111,4 +127,87 @@ static void M_Setup(OBJECT *const obj)
+    Drawbridge_GetSectorPositions(item, pos);
+    for (int32_t i = 0; i < pos->count; i++) {
+        MovableBlock_DropStack(
+            item->pos.y, *(const XYZ_32 *)Vector_Get(pos, i), item->room_num);
```
Any chance of making this function more generic? Like, creating a proxy, `Item_DropStack` that calls `MovableBlock_DropStack`.
I'm thinking in the future it might be desirable for this logic to also affect pickup items, for example.

walkawayy replied:
I will think on how to best implement this for drop and slide. Unsure for now.

EDIT: This will be done in the future in a later PR.

rr- commented on src/libtrx/game/objects/general/drawbridge.c:
```diff
@@ -111,4 +127,87 @@ static void M_Setup(OBJECT *const obj)
+    pos = nullptr;
+}
+
+static int32_t floor_div(int32_t v, int32_t d)
```
`M_FloorDiv` + static decl

rr- commented on src/libtrx/game/objects/general/drawbridge.c:
```diff
@@ -111,4 +127,87 @@ static void M_Setup(OBJECT *const obj)
+        break;
+    }
+    return rot_bounds;
+}
```
This is largery a repetition of `Item_RotateBounds`. Would be cool to see this extracted somewhere, maybe `math/` module considering how it's not inherently related to items.

walkawayy replied:
Talked about this in another comment.

EDIT: I will make the change.

rr- commented on src/libtrx/game/objects/general/drawbridge.c:
```diff
@@ -111,4 +127,87 @@ static void M_Setup(OBJECT *const obj)
+        for (int32_t sz = sz0; sz <= sz1; ++sz) {
+            if (sx == sx_orig && sz == sz_orig)
+                // Skip the original sector since it's WALL_L away.
+                continue;
```
Missing braces

rr- commented on src/libtrx/game/objects/general/drawbridge.c:
```diff
@@ -111,4 +127,87 @@ static void M_Setup(OBJECT *const obj)
+
+            XYZ_32 pos = { .x = sx * WALL_L + WALL_L / 2,
+                           .y = item->pos.y,
+                           .z = sz * WALL_L + WALL_L / 2 };
```
Would be cool to see this reformatted as
```c
            const XYZ_32 pos = {
                .x = sx * WALL_L + WALL_L / 2,
                .y = item->pos.y,
                .z = sz * WALL_L + WALL_L / 2,
            };
```
(trailing comma should trigger a reindent)

rr- commented on src/libtrx/game/objects/general/drawbridge.c:
```diff
@@ -111,4 +127,87 @@ static void M_Setup(OBJECT *const obj)
+                           .y = item->pos.y,
+                           .z = sz * WALL_L + WALL_L / 2 };
+            LOG_DEBUG("Add pos: %d %d %d", pos.x, pos.y, pos.z);
+            Vector_Add(sector_pos, (void *)&pos);
```
Unnecessary cast to `(void*)`

rr- commented on src/libtrx/game/objects/general/trapdoor.c:
```diff
@@ -103,14 +117,92 @@ static void M_Control(const int16_t item_num)
+static int32_t floor_div(int32_t v, int32_t d)
+{
+    return (v >= 0) ? v / d : -((-v + d - 1) / d);
+}
```
This actually appears more than once so I'd suggest moving it into `math/` as a public function, `Math_FloorDiv`. Args can be const BTW

rr- commented on src/libtrx/game/objects/general/trapdoor.c:
```diff
@@ -103,14 +117,92 @@ static void M_Control(const int16_t item_num)
+        break;
+    }
+    return rot_bounds;
+}
```
And just a marker to remember to deduplicate this function too.

rr- commented on src/libtrx/include/libtrx/game/objects/traps/movable_block.h:
```diff
@@ -4,5 +4,23 @@
+    uint16_t gravity_frames;
+    bool is_push_pull;
+    XYZ_32 pos;
+} MovableBlock_Info;
```
Wrong case – should be `MOVABLE_BLOCK_INFO`

rr- commented on src/libtrx/include/libtrx/game/objects/traps/movable_block.h:
```diff
@@ -4,5 +4,23 @@
+void MovableBlock_SetGravityFrames(ITEM *item, uint8_t frames);
+uint16_t MovableBlock_GetGravityFrames(const ITEM *item);
+void MovableBlock_SetPos(ITEM *item);
+XYZ_32 MovableBlock_GetPos(const ITEM *item);
```
I'd name these functions, `SetPos` and `GetPos`, something that stands out more, as currently it's unclear how `MOVABLE_BLOCK_INFO.pos` is different from `ITEM.pos` and how it should be used.

walkawayy replied:
For the movable blocks and sliding pillars, the `item->data`'s positions is sort of a last known, good, old position. It's needed for both to reposition their walkable without allocating more memory. I wasn't 100% sure what to call it either. I'll think on it.

walkawayy replied:
I've decided to go with `linked_sector_pos`. I think that's the best description. It's basically the last good, sector location.

rr- commented on src/libtrx/include/libtrx/game/rooms/utils.h:
```diff
@@ -4,3 +4,5 @@
 #define ROUND_TO_CLICK(V) ((V) & ~(STEP_L - 1))
 #define ROUND_TO_SECTOR(V) ((V) & ~(WALL_L - 1))
+#define ROUND_TO_CLICK_UP(V) (((V) + STEP_L / 2 - 1) & ~(STEP_L / 2 - 1))
+#define ROUND_TO_HALF_CLICK(V) ((V) & ~(STEP_L / 2 - 1))
```
Great to see you getting used to bitwise operations :D

rr- commented on src/tr1/game/savegame/savegame_bson.c:
```diff
@@ -625,6 +627,32 @@ static bool M_LoadItems(JSON_ARRAY *items_arr, uint16_t header_version)
+                        i);
+                    return false;
+                }
+                MovableBlock_Info *const data = (MovableBlock_Info *)item->data;
```
No need for the cast

rr- commented on src/tr2/game/objects/vars.c:
```diff
@@ -314,7 +314,6 @@ const GAME_OBJECT_ID g_BossObjects[] = {
 
 const GAME_OBJECT_ID g_PlaceholderObjects[] = {
     // clang-format off
-    O_BARTOLI,
```
I'm not sure if this change is intended, but if it is let's make it into a separate PR.

walkawayy replied:
Tbh idk how I did this lol. I reverted the deletion.

rr- commented on src/tr1/game/shell/common.c:
```diff
@@ -8,6 +8,10 @@
+#include <libtrx/game/items.h>
+#include <libtrx/game/music.h>
+#include <libtrx/game/option.h>
+#include <libtrx/game/ui.h>
```
I don't think these includes are necessary

rr- commented on src/libtrx/include/libtrx/game/objects/traps/movable_block.h:
```diff
@@ -4,5 +4,23 @@
+    bool reposition);
+
+typedef struct {
+    int16_t counter_rot[3];
```
I'm not sure I've seen any other property than the first member used – if that's the case let's make it a simple field.

walkawayy replied:
I think so too, but I will defer to @lahm86 on this one since this is his code.

lahm86 replied:
This array is needed because of how it's handled below.

https://github.com/LostArtefacts/TRX/blob/357534b35d8b011ab771d4fecdce19ca3ec2c907/src/tr1/game/objects/common.c#L221

It's similar for things - like the compass needle - that apply extra rotation. And `CREATURE` of course. I don't like that it's assuming layout for the latter; it'd be a sizeable refactor to improve it though.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -39,85 +33,192 @@ static bool M_IsValidFloorShiftState(const ITEM *const item)
+    stack = nullptr;
+}
+
+void MovableBlock_SlideStack(
```
Similar comment about proxying this into a `Item_SlideStack` or `Walkable_SlideStack` that calls this function in case we want to extend the sliding function to cover pickups.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -39,85 +33,192 @@ static bool M_IsValidFloorShiftState(const ITEM *const item)
 }
 
-static void M_ShiftGlobalFloorUp(void)
+static int32_t M_GetSectorIndex(int32_t x, int32_t divisor)
```
This function appears to be unused.

rr- reviewed (commented):
Mostly minor comments about cosmetic stuff, but I also left a few remarks on how we could avoid some duplication around utility methods, and move some of the stacking/sliding logic into the item module or walkable module.

I've mostly skimmed to highlight some classes of styling problems and I'll do a more thorough review once we're near merging, but ATM, overall this looks rather solid to me :relieved:

lahm86 commented on src/tr1/game/objects/traps/sliding_pillar.c:
```diff
@@ -1,91 +1,252 @@
-    } else {
-        Room_AlterFloorHeight(item, WALL_L * 2);
-    }
+    // Room_AlterFloorHeight(item, -WALL_L * 2);
```
The sliding pillar will need a `MovableBlock_UpdateBox` call in its initialise function to match OG. I've updated the test level with blockable sectors under 3 of the 4 sliding pillars.

As an aside, perhaps in the future after further merging `MovableBlock_UpdateBox` could be moved somewhere else common.

lahm86 reviewed (commented):
Everything in the test level is working really well. Just a few comments - couple of snags, but hopefully they are minor.
- ~~Killer pushblocks have stopped working~~ Edit: confirmed I had `immune` on, so my bad!
- Initial blocked box for sliding pillars (see below).
- The camera around sliding pillars seems normal to me (compare OG Qualopec room). Similar to the blocks, using TR2's camera gives better results, so it's just a quirk IMO.

I've also added an antipad (Natla texture) for the pushblocks in the pillar room if you want to dabble with the functionality of being able to reset their positions.

walkawayy committed:
math: add floor division function

walkawayy committed:
walkable: add functions to handle walkable items

walkawayy committed:
tr1/walkables: add and fix numerous interactions

- add the ability for movable blocks to move on trapdoors,
  drawbridges, bridges, sliding pillars, and falling blocks
- add the ability for stacks of movable blocks to fall and land
  on trapdoors, drawbridges, bridges, sliding pillars, and falling
  blocks
- add the ability for stacks of movable blocks to fall when
  on opened trapdoors and drawbridges
- fix various bugs with falling movable blocks

walkawayy commented:
Majorly cleaned up the code. I have a few outstanding questions:

* @lahm86 Do you have immune Lara cheat turned on? Deadly falling movable blocks seem to be working fine for me.
* Can we delay rewriting the sliding and dropping functions to more general functions for a future issue and PR?
* ~~This PR currently focuses only on TR1. Do we want to include a move to TRX and TR2 support (with lifts maybe) now or work on that in a future PR?~~ This may need to happen in this PR as current libtrx changes actually cause issues with TR2. Once we think TR1 support is wrapped up, I can move to libtrx/TR2 support.

rr- commented:
>Can we delay rewriting the sliding and dropping functions to more general functions for a future issue and PR?

Yes for sure.

walkawayy committed:
sound: rename SFX_T_REX_FOOTSTOMP/__FEET to SFX_T_REX_STOMP

walkawayy committed:
movable_block: move to libtrx and add TR2 walkables support

Resolves #2758.
Resolves #3784.
Resolves #3828.

- add the ability for movable blocks to move on trapdoors,
  drawbridges, bridges, lifts, and falling blocks
- add the ability for stacks of movable blocks to fall and land
  on trapdoors, drawbridges, bridges, lifts, and falling blocks
- add the ability for stacks of movable blocks to fall when on
  opened trapdoors and drawbridges
- add the ability for movable blocks to travel up and down lifts
- fix various bugs with falling movable blocks
- fix Lara hang climbing up a movable block used as a ladder piece

walkawayy committed:
libtrx: move falling block module to TRX

walkawayy committed:
movable_block: correct tunneling and room number

walkawayy commented:
Force pushed a change to fix the walkable initialization that was showing itself when loading the test level with the `-l` shortcut arg. The walkables are now properly initialized after object setup but before item initialization.

aredfan commented:
**Issues:**

- [x] If Lara interacts with a block while both are inside a lift, just before the lift starts moving, Lara will get left behind.  You'll need to be quite quick to get Lara to interact with a block in this scenario.
- [x] If the player save and reload while a lift is moving with a block inside or top of it, the block will get left behind.
- [x] It's possible for lifts to leave blocks behind, this bug only happens on the ground floor.

> 1. go to the  ground floor and pull the switch so the elevator is at the bottom
> 2. save and reload
> 3. push the block into the elevator
> 4. pull the switch and watch the block get left behind

- [x]  It's possible Lara cannot pull blocks backwards away from the lift's top, see screenshot

> 1. go to room 55 and put the block into the corner at the top of the lift
> 2. pull the switch to lower the lift, then pull the switch again to raise the lift
> 3. try having Lara pull the block backwards
> 
> <details><summary>Screenshot</summary>
> 
> ![20250830_150103_Test_Level](https://github.com/user-attachments/assets/add24580-adc8-4c8c-a2b7-679a7192de91)
> 
> </details>

- [x] If Lara runs into a block while it's under a trapdoor, she'll land on top of it. It's a SG bug. Here's the SG which is for the test level. [save_tr2_custom_00.zip](https://github.com/user-attachments/files/22058825/save_tr2_custom_00.zip)

walkawayy commented:
> **Issues:**
> 
> * [x]  If Lara interacts with a block while both are inside a lift, just before the lift starts moving, Lara will get left behind.  You'll need to be quite quick to get Lara to interact with a block in this scenario.
> * [x]  If the player save and reload while a lift is moving with a block inside or top of it, the block will get left behind.
> * [x]  It's possible for lifts to leave blocks behind, this bug only happens on the ground floor.
> 
> 1. go to the  ground floor and pull the switch so the elevator is at the bottom
> 2. save and reload
> 3. push the block into the elevator
> 4. pull the switch and watch the block get left behind
> 
> * [x]   It's possible Lara cannot pull blocks backwards away from the lift's top, see screenshot
> 
> 1. go to room 55 and put the block into the corner at the top of the lift
> 2. pull the switch to lower the lift, then pull the switch again to raise the lift
> 3. try having Lara pull the block backwards
> 
> * [x]  If Lara runs into a block while it's under a trapdoor, she'll land on top of it. It's a SG bug. Here's the SG which is for the test level. [save_tr2_custom_00.zip](https://github.com/user-attachments/files/22058825/save_tr2_custom_00.zip)



I pushed a work in progress commit that I think should fix all the bugs listed except the last one. I couldn't reproduce it but it's possible a different bug fix I made also fixed it?

aredfan commented:
Thank you @walkawayy. 👍 

I've come across the following issues in the sliding pillars room.

- [x] Issue 1

> 1. `/tp 32 -2 49`
> 2. turn around and push the block into the corner
> 3. save and reload
> 4. Lara is unable to pull the block back, and there's an odd collision issue if Lara runs into the block

- [x] Issue 2

> 1. drop the blocks onto the sliding pillars
> 2. save and reload
> 3. move the sliding pillars and observe the blocks getting left behind

- [x] Issue 3

> 1. drop the blocks onto the sliding pillars
> 2. move the sliding pillars, then quickly save and reload while the pillar is midway
> 3. observe the blocks getting left behind

- [x] Issue 4

> 1. have the blocks fall to the ground (not onto the pillars)
> 2. move one of the blocks 1 tile in any direction
> 3. save and reload
> 4. this causes an odd collision issue with the block

rr- commented on docs/tr1/CHANGELOG.md:
```diff
@@ -99,8 +103,15 @@ Showcase: https://www.youtube.com/watch?v=iV8G9lhxVQ8
+    - added the ability for stacks of movable blocks to fall and land on trapdoors, drawbridges, bridges, sliding pillars, and falling blocks
+    - added the ability for stacks of movable blocks to fall when on opened trapdoors or drawbridges
+    - fixed various bugs with falling movable blocks
+>>>>>>> e6d20a531 (falling_block: drop movable blocks when trap falls)
```
Conflicts markers. Plus added to the wrong version. Let's make sure it's at the top of the list.

rr- commented on docs/tr2/CHANGELOG.md:
```diff
@@ -2,6 +2,13 @@
 - fixed missing shader and configuration files from MacOS releases (#3870, regression from 1.4)
 - fixed zero byte at the end of config files (#3875, regression from 1.4)
 - fixed stacked sprites flickering (#3872, regression from 1.4)
+- added the ability for falling movable blocks to kill Lara outright if one lands directly on her (#3784)
```
Added to the wrong version; let's make sure it's at the top of the list.

rr- commented on docs/tr2/IMPROVEMENTS.md:
```diff
@@ -137,6 +138,12 @@
+    - added the ability for stacks of movable blocks to fall when on trapdoors, drawbridges, or falling blocks
+    - added the ability for movable blocks to travel up and down lifts
+    - fixed various bugs with falling movable blocks
+- fixed Lara hang climbing up a movable block used as a ladder piece
```
Missing similar entries in tr1/IMPROVEMENTS.md.

walkawayy replied:
Lifts and ladders shouldn't apply to TR1 right?

rr- replied:
Lifts no ladders I think yes? Or maybe this work hasn't been done yet. I don't remember off the top of my head.
Unless you mean that the pushblock-related improvements are already in tr1/IMPROVEMENTS.md which is why there's no diff in that file in which case it's totally fine.

rr- commented on src/libtrx/game/game_buf.c:
```diff
@@ -3,12 +3,12 @@
 {
     for (int32_t i = 0; i < GBUF_NUM_MALLOC_TYPES; i++) {
-        m_Allocator[i].default_chunk_size = 2048;
+        m_Allocator[i].default_chunk_size = 1024;
```
Let's avoid changes to this module

rr- commented on src/libtrx/game/game_flow/reader_tr2.def.c:
```diff
@@ -135,6 +135,8 @@ static void M_LoadRoot(const M_CONTEXT *const ctx, JSON_OBJECT *const obj)
 
     ctx->gf->secret_track = JSON_ObjectGetInt(obj, "secret_track", MX_INACTIVE);
+    ctx->gf->enable_killer_pushblocks =
+        JSON_ObjectGetBool(obj, "enable_killer_pushblocks", true);
```
Now that both TR1 and TR2 understands this bool, we can move reading of it to `M_LoadCommonRoot`

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,130 @@
+            return;
+        }
+        walkables = &(*walkables)->next;
+    }
```
This iteration can be simplified here and in other places:
```c
    const WALKABLE *walkable = sector->walkable;
    while (walkable != nullptr) {
        if (walkable->item_num == item_num) {
            return;
        }
        walkable = walkable->next;
    }
```

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,130 @@
+
+        SECTOR *sec = room->sectors;
+        for (int32_t z = 0; z < z_max; z++)
+            for (int32_t x = 0; x < x_max; x++, sec++) {
```
Let's make sure the `sec++` accumulator is more visible by putting it under the `sec->walkable = nullptr` assignment.

rr- commented on src/libtrx/game/objects/general/bridge_flat.c:
```diff
@@ -55,10 +55,15 @@ static void M_Setup(OBJECT *const obj)
 {
+    const ITEM *const item = Item_Get(item_num);
+    const OBJECT *const obj = Object_Get(item->object_id);
+    const ANIM_FRAME *const frame = Object_GetAnim(obj, 0)->frame_ptr;
```
It seems we are not using these new variables.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+
+void MovableBlock_UpdateBox(const ITEM *const item, const bool blocked)
+{
+    // TODO Might be other cases...
```
Can we elaborate or remove this TODO?

rr- commented on src/libtrx/game/rooms/geometry.c:
```diff
@@ -164,6 +164,46 @@ SECTOR *Room_GetSector(
+            room = Room_Get(*room_num);
+            sector = Room_GetWorldSector(room, x, z);
+            if (y < sector->floor.height)
+                break;
```
Missing { }

rr- commented on src/libtrx/game/rooms/geometry.c:
```diff
@@ -164,6 +164,46 @@ SECTOR *Room_GetSector(
+            room = Room_Get(*room_num);
+            sector = Room_GetWorldSector(room, x, z);
+            if (y >= sector->ceiling.height)
+                break;
```
Missing { }

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+{
+    item->rot.y = rot_y;
+    MOVABLE_BLOCK_INFO *const data = item->data;
+    data->counter_rot[0] = data->original_rot - rot_y;
```
I know we've talked about this but I don't see any usages of `counter_rot` other than this assignment?

walkawayy replied:
Linking this to Lahm's answer from the previous code review: https://github.com/LostArtefacts/TRX/pull/3671#discussion_r2262102693

rr- replied:
Ya I still don't get it though.
Oh wait… Oh my god. Because this is part of `item->data`, and some code casts `item->data` as `XYZ_16*`, and we happen to trigger those code paths with movable blocks but use a different struct for its private data… so it's important for this to be the FIRST entry. Ugh!
Can we put a big loud comment about why this exists? It's so easy to see no usages and just bulldoze it away as it is now.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+static bool M_IsPushPull(const ITEM *const item)
+{
+    const MOVABLE_BLOCK_INFO *const data = item->data;
+    return data ? data->is_push_pull : false;
```
Let's use explicit `data != nullptr` here and in other places.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+    const SECTOR *sector = Room_GetSector(x, y, z, &room_num);
+    COLL_INFO coll;
+    coll.quadrant = quadrant;
+    coll.radius = 500;
```
Let's use compact assigment like so: `coll = { .quadrant = quadrant, .radius = 500 };`.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+    COLL_INFO coll = {
+        .quadrant = quadrant,
+        .radius = 500,
+        0,
```
No need for `0` in C2x. And if we remove trailing comma we can fit this in a single line.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+    const ITEM *const item, const int32_t x, const int32_t z)
+{
+    int32_t dx = x - item->pos.x;
+    int32_t dz = z - item->pos.z;
```
Can be `const`

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+                box->overlap_index |= BOX_BLOCKED;
+            } else {
+                box->overlap_index &= ~BOX_BLOCKED;
+            }
```
We can use the `TOGGLE_BIT` macro to safe on the LOC a bit.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+REGISTER_OBJECT(O_MOVABLE_BLOCK_1, M_Setup)
+REGISTER_OBJECT(O_MOVABLE_BLOCK_2, M_Setup)
+REGISTER_OBJECT(O_MOVABLE_BLOCK_3, M_Setup)
+REGISTER_OBJECT(O_MOVABLE_BLOCK_4, M_Setup)
```
All in all this module has grown quite a lot. I wonder if there's any way to keep it smaller?

rr- commented on src/libtrx/game/items/walkable.c:
```diff
@@ -0,0 +1,130 @@
+        target.pos.y, target.pos.z);
+}
+
+void Walkable_Shutdown(void)
```
Let's call this `Walkable_Unload` or `Walkable_Reset` as `Shutdown` / `Init` habitually refers to program startup and termination, whereas here we call this mid-runtime on level load.

rr- commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,961 @@
+static void M_AddWalkable(int16_t item_num);
+static void M_GetStack(
+    VECTOR *stack, XYZ_32 stack_pos, int32_t stack_height, int32_t step_y,
+    int16_t room_num);
```
Let's consider adding line breaks to group similar functions together.
Also for some of the trickier logic it might be useful to put forth some docstrings.

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -1074,6 +1143,21 @@ static bool M_LoadItems(JSON_ARRAY *const items_arr)
+                        "item_num: %d; is_moving: %d", i, data->is_moving);
+                    for (int32_t j = 0; j < LIFT_NUM_SECTORS; j++) {
+                        char pos_key[50];
+                        snprintf(pos_key, 50, "linked_%d", j);
```
Can use `String_FormatStatic` for this.

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -1074,6 +1143,21 @@ static bool M_LoadItems(JSON_ARRAY *const items_arr)
+                        char pos_key[50];
+                        snprintf(pos_key, 50, "linked_%d", j);
+                        LOAD_XYZ(data_obj, pos_key, data->linked[j]);
+                        LOG_DEBUG(
```
I'm not sure we need this log

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -978,6 +1015,38 @@ static bool M_LoadItems(JSON_ARRAY *const items_arr)
+                LOAD_XYZ(data_obj, "linked", data->linked);
+                LOG_DEBUG(
+                    "i: %d; xyz: %d %d %d; room_num: %d", i, data->linked.pos.x,
+                    data->linked.pos.y, data->linked.pos.z, item->room_num);
```
I'm not sure we need this log

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -1074,6 +1143,21 @@ static bool M_LoadItems(JSON_ARRAY *const items_arr)
+                    data->is_moving = JSON_ObjectGetBool(
+                        data_obj, "is_moving", data->is_moving);
+                    LOG_DEBUG(
+                        "item_num: %d; is_moving: %d", i, data->is_moving);
```
I'm not sure we need this log

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -811,10 +838,20 @@ static JSON_ARRAY *M_DumpItems(void)
+                DUMP_XYZ(data_obj, pos_key, data->linked[j]);
+                LOG_DEBUG(
+                    "j: %d; xyz: %d %d %d", j, data->linked[j].pos.x,
+                    data->linked[j].pos.y, data->linked[j].pos.z);
```
I'm not sure we need this log

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -811,10 +838,20 @@ static JSON_ARRAY *M_DumpItems(void)
             JSON_ObjectAppendInt(data_obj, "start_height", data->start_height);
             JSON_ObjectAppendInt(data_obj, "wait_time", data->wait_time);
+            JSON_ObjectAppendBool(data_obj, "is_moving", data->is_moving);
+            LOG_DEBUG("item_num: %d; is_moving: %d", i, data->is_moving);
```
I'm not sure we need this log

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -811,10 +838,20 @@ static JSON_ARRAY *M_DumpItems(void)
+            LOG_DEBUG("item_num: %d; is_moving: %d", i, data->is_moving);
+            for (int32_t j = 0; j < LIFT_NUM_SECTORS; j++) {
+                char pos_key[50];
+                snprintf(pos_key, 50, "linked_%d", j);
```
Can use `String_FormatStatic` for this

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -751,6 +753,31 @@ static JSON_ARRAY *M_DumpItems(void)
+                DUMP_XYZ(data_obj, "linked", data->linked);
+                LOG_DEBUG(
+                    "i: %d; xyz: %d %d %d", i, data->linked.pos.x,
+                    data->linked.pos.y, data->linked.pos.z);
```
I'm not sure we need this log

rr- commented on src/tr2/game/savegame/savegame_bson.c:
```diff
@@ -751,6 +753,31 @@ static JSON_ARRAY *M_DumpItems(void)
             }
+
+            if (Object_IsType(item->object_id, g_MovableBlockObjects)
+                && item->data) {
```
`item->data != nullptr`

rr- commented on src/tr2/game/objects/general/lift.h:
```diff
@@ -1,8 +1,16 @@
 #include <stdint.h>
 
+#define LIFT_NUM_FLOOR_SECTORS 4
+#define LIFT_NUM_SECTORS 8
```
Eventually it would be nice for the lift to stop once it touches the floor rather than going for a fixed number of sectors. I'm not sure how difficult that would be to integrate with.

rr- commented on src/tr1/game/savegame/savegame_bson.c:
```diff
@@ -625,6 +645,58 @@ static bool M_LoadItems(JSON_ARRAY *items_arr, uint16_t header_version)
+                data->is_forced_moving = JSON_ObjectGetBool(
+                    data_obj, "is_forced_moving", data->is_forced_moving);
+                LOAD_XYZ(data_obj, "linked", data->linked);
+                LOG_DEBUG(
```
I'm not sure we need this log.

rr- commented on src/tr1/game/savegame/savegame_bson.c:
```diff
@@ -625,6 +645,58 @@ static bool M_LoadItems(JSON_ARRAY *items_arr, uint16_t header_version)
+                LOG_DEBUG(
+                    "i: %d; xyz: %d %d %d; room_num: %d", i, data->linked.pos.x,
+                    data->linked.pos.y, data->linked.pos.z,
+                    data->linked.room_num);
```
I'm not sure we need this log.

rr- commented on src/tr1/game/savegame/savegame_bson.c:
```diff
@@ -1184,6 +1256,35 @@ static JSON_ARRAY *M_DumpItems(void)
             }
+
+            if (Object_IsType(item->object_id, g_MovableBlockObjects)
+                && item->data) {
```
`item->data != nullptr`

rr- commented on src/tr1/game/savegame/savegame_bson.c:
```diff
@@ -34,6 +37,23 @@ typedef struct {
+        value.x = JSON_ObjectGetInt(sub_obj, "x", value.x);                    \
+        value.y = JSON_ObjectGetInt(sub_obj, "y", value.y);                    \
+        value.z = JSON_ObjectGetInt(sub_obj, "z", value.z);                    \
+    } while (0)
```
Nice!

walkawayy replied:
Stolen from Lahm's TR2 savegame_bson.c xd. Any suggestions on where we want to put these common macros?

rr- replied:
Sorry I didn't mean to be sarcastic I hope I wasn't taken this way. I genuinely like these macros. I'd say let's keep it as a dupe for now until we tackle merging the SG modules which is a task on its own.

walkawayy replied:
Lol you're fine. I was just making a joke since you praised me for it but all I did was copy Lahm's macro from TR2's savegame code.

rr- commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,49 +183,189 @@ static void M_Initialise(const int16_t item_num)
+    M_GetSectorPositions(item, positions);
+    for (int32_t i = 0; i < positions->count; i++) {
+        GAME_VECTOR linked = { .pos = *(const XYZ_32 *)Vector_Get(positions, i),
+                               .room_num = item->room_num };
```
Let's add a trailing comma prior to } for this to format more nicely.

rr- commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,49 +183,189 @@ static void M_Initialise(const int16_t item_num)
+        // LIFT_INFO *const lift_data = item->data;
+        // const XYZ_32 linked_sector_pos = lift_data->linked_sector_pos;
+        // Walkable_Reposition(item_num, item->pos, linked_sector_pos);
+        // lift_data->linked_sector_pos = item->pos;
```
Are these TODOs needed or can we confirm this can go away?

rr- commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,49 +183,189 @@ static void M_Initialise(const int16_t item_num)
+                    item->pos.y, item->room_num, true);
+                // Don't reposition because item->pos links to a single sector.
+                lift_data->linked[i].pos.y = item->pos.y;
+                LOG_DEBUG(
```
I'm not sure we need this log. Consider using `LOG_TRACE` which is a no-op unless the developer enables it.

rr- commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,49 +183,189 @@ static void M_Initialise(const int16_t item_num)
+                // Don't reposition because item->pos links to a single sector.
+                lift_data->linked[i].pos.y = item->pos.y - LIFT_HEIGHT;
+                LOG_DEBUG(
+                    "Set linked[i].pos.y: %d", lift_data->linked[i].pos.y);
```
I'm not sure we need this log. Consider using `LOG_TRACE` which is a no-op unless the developer enables it.

rr- commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,49 +183,189 @@ static void M_Initialise(const int16_t item_num)
+            ? (delta < LIFT_SHIFT ? delta : LIFT_SHIFT)
+            : (delta > -LIFT_SHIFT ? delta : -LIFT_SHIFT);
+        item->pos.y += step;
+        LOG_DEBUG("item->pos.y: %d", item->pos.y);
```
I'm not sure we need this log. Consider using `LOG_TRACE` which is a no-op unless the developer enables it.

rr- commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,49 +183,189 @@ static void M_Initialise(const int16_t item_num)
+    }
+
+    // Collect a 2×2 footprint that lines up with the shaft tiles.
+    for (int ix = 0; ix < 2; ++ix) {
```
We favor postincrementation over preincrementation in loop accumulators (this line and the one below)

rr- commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,49 +183,189 @@ static void M_Initialise(const int16_t item_num)
+
+    // Collect a 2×2 footprint that lines up with the shaft ceiling tiles.
+    for (int ix = 0; ix < 2; ++ix) {
+        for (int iz = 0; iz < 2; ++iz) {
```
Ditto

rr- commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,49 +183,189 @@ static void M_Initialise(const int16_t item_num)
+                .y = item->pos.y - LIFT_HEIGHT,
+                .z = sz * WALL_L + WALL_L / 2,
+            };
+            Vector_Add(sector_pos, &pos);
```
Can these loops be merged together, or does the order inside `sector_pos` matter and we can't interleave positions like that?

aredfan commented:
**Sliding pillars (part 2)**

- [x] Issue 1

> 1. `/tp 32 -2 47`
> 2. push the block onto the pillar, and when the pillar is moving, save and reload
> 3. observe the block getting left behind

- [x] Issue 2

> 1. `/tp 34 0 53`
> 2. step on the nearest pad
> 3. release all blocks
> 4. step on the nearest antipad
> 5. step on the corresponding pad, and when the pillar is moving, save and reload
> 6. observe the block getting left behind

**Lifts (part 2)**

- [ ] Issue 3 (present in develop too)

> 1. move the block next to a lift
> 2. have Lara climb the block, then walk onto the lift
> 3. Lara will clip into the lift
> <details><summary>Screenshot</summary>
> 
> ![20250911_195259_Test_Level](https://github.com/user-attachments/assets/fcb034cd-ee1b-4308-8a4c-e58258de8bd6)
> 
> </details>

In the PR build, if Lara clips into the lift before it's activated, she can walk out through the closed doors.  In develop, Lara is unable to walk out through the close doors.  I'm not sure if it's a bug worth fixing but thought I'd mention it.

walkawayy commented:
Pushed a fix for the sliding pillars and movable blocks loading which should fix those issues. Regarding lifts, I did notice Lara can clip into that lift from the block which is indeed an OG bug. I will try to fix the bug where Lara can exit the lift early in the PR build.

I think other than that, all other lift issues should be a continued in a new PR linked to OG issues. There is another OG issue that can happen where Lara falls out of the lift if she runs hard enough into a wall to trigger her wall splat animation when the lift is moving up. Lara will splat and fall if the lift is traveling down, but does not fall out of the lift.

![20250911_154807_Test_Level](https://github.com/user-attachments/assets/87e5d76a-cf1a-4e27-a013-1dd9d94369b5)

![20250911_154807_Test_Level_2](https://github.com/user-attachments/assets/0c8c2eda-4a91-4d74-baff-77e0de6dc073)

walkawayy committed:
tr1/savegame: add dump and load XYZ macros

lahm86 commented on src/libtrx/game/objects/general/trapdoor.c:
```diff
@@ -88,13 +100,26 @@ static bool M_IsItemOnTop(
+        Walkable_Add(item_num, *(const XYZ_32 *)Vector_Get(positions, i));
+    }
+    Vector_Free(positions);
+    positions = nullptr;
```
I think this line is redundant.

lahm86 commented on src/libtrx/game/objects/general/trapdoor.c:
```diff
@@ -112,6 +138,78 @@ static void M_Control(const int16_t item_num)
+            *(const XYZ_32 *)Vector_Get(positions, i), item->room_num);
+    }
+    Vector_Free(positions);
+    positions = nullptr;
```
Redundant line here too.

lahm86 commented on src/libtrx/game/objects/traps/falling_block.c:
```diff
@@ -70,6 +88,9 @@ static void M_Control(const int16_t item_num)
 
     case TRAP_WORKING:
         if (item->goal_anim_state != TRAP_FINISHED) {
+            if (item->gravity == false) {
```
I think we prefer `!item->gravity` rather than this when the type is a bool?

rr- replied:
Correct `!` is kind of reserved for normal booleans.

walkawayy replied:
Ok I wasn't sure since we switched to explicit `data != nullptr`

lahm86 commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,963 @@
-    qsort(m_SortedBlocks, m_BlockCount, sizeof(int16_t), M_CompareBlock);
-    M_ShiftGlobalFloorUp();
+    LARA_INFO *const lara = Lara_GetLaraInfo();
+    DIRECTION quadrant = ((uint16_t)lara_item->rot.y + DEG_45) / DEG_90;
```
`const DIRECTION...`

lahm86 commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,963 @@
+            item->pos.x, item->pos.y - STEP_L, item->pos.z, &sector_room_num);
+        Item_UpdateRoom(item_num, sector_room_num);
+        if (reposition) {
+            GAME_VECTOR target = { .pos = item->pos,
```
`const GAME_VECTOR...`

walkawayy replied:
Fixed all these and added the trailing comma after the `room_num` init so they format nicely.

lahm86 commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,963 @@
+    }
+
+    Vector_Free(stack);
+    stack = nullptr;
```
Redundant

lahm86 commented on src/libtrx/game/objects/traps/movable_block.c:
```diff
@@ -1,123 +1,963 @@
+        item->pos.x = dest_item->pos.x;
+        item->pos.z = dest_item->pos.z;
+        int16_t sector_room_num = dest_item->room_num;
+        SECTOR *sector = Room_GetSector(
```
I think the assignment can be removed as this is only used to get the room number, if I understand correctly.

lahm86 commented on src/libtrx/game/rooms/geometry.c:
```diff
@@ -164,6 +164,48 @@ SECTOR *Room_GetSector(
+    }
+
+    // Check if on a walkable.
+    bool skip_pit = false;
```
I think this can be declared on line 182 and made `const`

lahm86 commented on src/tr1/game/objects/traps/sliding_pillar.c:
```diff
@@ -1,91 +1,228 @@
+static bool M_IsItemOnTop(
+    const ITEM *const item, const int32_t x, const int32_t z)
+{
+    int32_t dx = x - item->pos.x;
```
`const`, both lines

lahm86 commented on src/tr1/game/objects/traps/sliding_pillar.c:
```diff
@@ -1,91 +1,228 @@
 
     Item_Animate(item);
-
+    int16_t old_room_num = item->room_num;
```
I don't think this var is used.

lahm86 commented on src/tr1/game/objects/traps/sliding_pillar.c:
```diff
@@ -1,91 +1,228 @@
         item->pos.x += WALL_L / 2;
         item->pos.z &= -WALL_L;
         item->pos.z += WALL_L / 2;
+        GAME_VECTOR target = {
```
`const `

lahm86 commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -166,6 +181,19 @@ static void M_Initialise(const int16_t item_num)
+        lift_data->linked[i] = linked;
+    }
+    Vector_Free(positions);
+    positions = nullptr;
```
Redundant.

lahm86 commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -174,41 +202,148 @@ static void M_Control(const int16_t item_num)
+        item->goal_anim_state = LIFT_STATE_DOOR_OPEN;
+        lift_data->wait_time = 0;
+        if (lift_data->is_moving) {
+            for (int i = 0; i < LIFT_NUM_FLOOR_SECTORS; i++) {
```
`int32_t`, a few similar cases in this module.

lahm86 commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -174,41 +202,148 @@ static void M_Control(const int16_t item_num)
+        Walkable_Add(item_num, *(const XYZ_32 *)Vector_Get(positions, i));
+    }
+    Vector_Free(positions);
+    positions = nullptr;
```
Redundant

walkawayy replied:
the `nullptr` assignments? I only need to call `Vector_Free`?  Ok I'll clean those up. There's some other code I looked at that did the same which I lifted it from.

`text.c`:
```c
        Vector_Free(s->page_content);
        s->page_content = nullptr;
```

`handlers_language.c`:
```c
        Vector_Free(m_Languages);
        m_Languages = nullptr;
```

lahm86 replied:
Yeah, in the cases I noticed above, the vars have function scope only so the pointer is gone once the function ends essentially. In those other cases you listed, it avoids dangling pointers for module or heap vars that don't go out of scope.

lahm86 reviewed (commented):
This is all looking great 👏 I've had a play with the test levels again and it's looking good to me.
Some super minor nitpicks below. The only other thing is I think the file `src/tr1/game/objects/traps/movable_block.c` can be deleted - it's unlinked in `meson.build` but still in the repo (apologies if you're already aware).

walkawayy commented:
Force pushed the corrected commits from Lahm's code review comments.
EDIT: and rebased the latest develop

walkawayy commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -77,43 +89,47 @@ static void M_FloorCeiling(
+
+    const bool lara_inside_lift = (lara_item->pos.y < lift_bottom) &&
+                                  (lara_item->pos.y > lift_ceiling);
+    // clang-format on
```
One additional cosmetic change I did make was clean up these names which I hope people agree are easier to understand. Here's a picture to demonstrate:

<img width="3840" height="2160" alt="image" src="https://github.com/user-attachments/assets/b095d1d2-b3db-4b5c-a178-a5b19609288e" />

walkawayy commented on src/tr2/game/objects/general/lift.c:
```diff
@@ -77,43 +89,47 @@ static void M_FloorCeiling(
     const int32_t lift_floor = item->pos.y;
-    const int32_t lift_ceiling = item->pos.y - LIFT_HEIGHT;
+    const int32_t lift_ceiling = item->pos.y - LIFT_HEIGHT + STEP_L;
+    const int32_t lift_top = item->pos.y - LIFT_HEIGHT;
```
One additional cosmetic change I did make was clean up these names which I hope people agree are easier to understand. Here's a picture to demonstrate:

<img width="3840" height="2160" alt="image" src="https://github.com/user-attachments/assets/b095d1d2-b3db-4b5c-a178-a5b19609288e" />
